### ISSUE TITLE
vim opens two windows with zathura

### ISSUE THREAD

**Issue by jayeshbadwaik** [lervag/vimtex#1017]
### Explain the issue
I have setup vim to work with zathura as described below. My local leader is `\`. Whenever I start open my document and start compilation mode, I get a window in zathura (call it window 1). Then, next, when I start a forward search, I get another window (call it window 2). I can do reverse search from the from window 1, and forward search to window 2, but I cannot do reverse search from window 2 and forward search from window 1.  

Let me know if you need more information to solve this problem.  I am running on the latest update of ArchLinux. 

```
$ xdotool --version 
xdotool version 3.20160805.1
```

### Minimal working example

```tex
\documentclass{minimal}
\begin{document}

Hello World!

\end{document}
```

### Minimal vimrc file
```vim
set nocompatible
let &rtp  = '~/.vim/plugged/vimtex,' . &rtp
let &rtp .= ',~/.vim/plugged/vimtex/after'
filetype plugin indent on
syntax enable

" VimTeX Specific Config
let g:tex_flavor = 'latex'
"" Local Leader
:let maplocalleader = "\\"
:let g:vimtex_view_method='zathura'
```

**Comment by lervag** [comment]
First: The default local leader is already `\`, so no reason to specify that explicitly.

So: I can't reproduce this. I'm also on Arch Linux, and everything works as expected here.

Could you please use the minimal setup you provided:

1. Open Vim (`vim -u minivimrc minimal.tex`)
2. Start compilation with `\ll`
3. Do `\li` to show some vimtex info, please print the output here.
4. Do `\lv` to forward search: if your problem persists this should open a new zathura window.
5. Repeat step 3.

Also, do you have a `.latexmkrc` file?

**Comment by jayeshbadwaik** [comment]
I do not have a .latexmkrc file.

Output after step 3:  
[step3.log](https://github.com/lervag/vimtex/files/1529579/step3.log)

Output after step 5:
[step5.log](https://github.com/lervag/vimtex/files/1529580/step5.log)

The diff between the two output is below:
    $ diff step3.log step5.log 
    31c31,32
    <     xwin id: 0
    ---
    >     xwin id: 46137347
    >     cmd_forward_search: zathura --synctex-forward 1:1:'/media/work/xlab/latex /minimal.tex' '/media/work/xlab/latex/minimal.pdf'

**Comment by lervag** [comment]
Great, the logs explain the issue, as I hoped. See the fifth line:

> Servername: undefined (vim started without --servername)

In order for latexmk to communicate with Vim to set the `xwin id` after compilation, when `zathura` is started from latexmk, we need to have a defined servername.

So, either you always start Vim like this: `vim --servername SOMENAME ...`, or you follow [this Vim Wiki tip](http://vim.wikia.com/wiki/Enable_servername_capability_in_vim/xterm) to make this automatic

**Comment by jayeshbadwaik** [comment]
So, I ran the command with servername and I get the same issue. Logs are attached. Same goes for GVim: 

[step5.log](https://github.com/lervag/vimtex/files/1533594/step5.log)
[step3.log](https://github.com/lervag/vimtex/files/1533595/step3.log)

**Comment by lervag** [comment]
I use fluxbox, but I don't think the desktop environment should have anything to do with it.

Can you now repeat steps 1-2, then from a terminal do:

```sh
xdotool search --name minimal.pdf
```

This should report an ID similar to what you get in step 5. I think this is where things are failing right now. Could you also try `xdotool search --class Zathura`?

As before: Do you have a `.latexmkrc` file? Also, which version of `latexmk` are you on?

**Comment by jayeshbadwaik** [comment]
No, I do not have a .latexmkrc file. 

```
→ latexmk --version

Latexmk, John Collins, 19 Jan. 2017. Version 4.52c
```

````
→ xdotool search --name minimal.pdf
106954755

→ xdotool search --class Zathura   
106954753
106954755
````

**Comment by jayeshbadwaik** [comment]
Hi, so  have discovered the problem. The problem is symlinks. 

Suppose my folder is `~/xlab/latex` with the file `~/xlab/latex/minimal.tex`  where `~/xlab~  is symlinked to `/media/work/jayesh/xlab`, then I face the problem of two windows which I have described. 

However, if the folder is not symlinked, then there is no issue and forward-backward search works correctly.

**Comment by lervag** [comment]
I see. I've sort of reproduced this in the case when I try to do `\lv` immediately after latexmk has started the viewer. I've made some test files as you can see in the thread, where I test for different timings. It takes ~0.5 second after the compilation has started until the viewer opens and until the viewer ID can be captured.

**Comment by jayeshbadwaik** [comment]
So, I took your test case and ran it on my computer, and I went upto 10 seconds (10000m) and even then the test is failing.

**Comment by lervag** [comment]
Strange. Note, though, that I specifically used `nvim`, so you have to change to `vim` if you don't use neovim. I now see that the test script does not work as expected for vim, I'll look into it.

**Comment by jayeshbadwaik** [comment]
I have setup for boths vim and neovim, so I used neovim, since you use neovim as well.

**Comment by jayeshbadwaik** [comment]
This seems a very subtle issue somewhere, and I can understand that since I am the one who is able to reproduce it and you are not, it is difficult for you to work on it. 

So, I would like to understand what vimtex does in order to figure out what is going on, and may be I can pinpoint the issue. However, I do not know anything about vimtex internals, so, if you don't mind, it would be nice if you could just tell me the list of files I should look at for this particular codeflow. I will look at it and try to find out where exactly is the issue to my best ability. 

Thank you.

**Comment by lervag** [comment]
Ok, so, some pointers:

- Inside Vim and neovim, paths are resolved. When you do `\li`, you will see that the paths to the tex files are the actual paths, not the symlinked paths. One may still see the "actual" path with e.g. `echo system('pwd -L')`.
- In this case, the first zathura window is actually opened by latexmk, not by vimtex. The current problem is that we need to detect the zathura window, which should happen from within the callback functions. The callgraph is something like `vimtex#compiler#callback -> vimtex#view#zathura#latexmk_callback`, although the latter function is really `b:vimtex.viewer.latexmk_callback`. Note that `one#two#name()` means the function is defined in `autoload/one/two.vim`.
- Other relevant files are `autoload/vimtex/view/common.vim`, which defines the `xwin_exists()` function.

I'm pushing an update to the test files where things should run also with vim. Still, based on your last comment it seems the problem is not related to vim/neovim, and so the test should be relevant also with neovim. When you are looking at the issue, I propose that you use a sleep in the order of 600-1000 ms, because this _should_ work.

**Comment by jayeshbadwaik** [comment]
Thanks. I will look at this over the weekend and see what i can find.

**Comment by lervag** [comment]
Note: I think I can reproduce the problem with Vim. That is, the test always fails for Vim, but it does not fail with nvim.

**Comment by lervag** [comment]
Ok, no, I don't reproduce. I only forgot about servername for Vim. And the servername is only available after `VimEnter` autocmd, so the test script needs some adjustments.

**Comment by lervag** [comment]
Ok, so with the last version of the test script, everything seems to work with both vim and neovim.

Just a side note: You may use the option `g:vimtex_view_automatic` to disable the automatic opening of the viewer. Also note, for capturing the xwin_id from Vim, you need to use the `--servername` option. Without it, vimtex will not be able to use the callbacks that sets the variable.

**Comment by jayeshbadwaik** [comment]
Cool. Thanks. I will keep that in mind.

**Comment by jayeshbadwaik** [comment]
Okay, so I guess there are some inconsistencies between using full path and relative path in the viewers which causes this issue.  

[patch.txt](https://github.com/lervag/vimtex/files/1540906/patch.txt)

The above patch does the following for, it automatically opens a preview window (call it `A`) and then after almost a second, opens another preview window (call it `B`). Forward and backward search works correctly with `B`. 

So, for now, I disable the automatic viewer `g:vimtex_view_automatic` whence the first window (A) does not appear at all and the second window works correctly. This is the state as of now, however, I have not yet pinpointed why the two different windows open and may be it is because there is a inconsistency in the callback from latexmk and view functions. I will check this when I have time.

**Comment by lervag** [comment]
Thanks for the patch. However, I don't think this works quite as you expect. When you apply this patch, what happens is that the callback from latexmk will definately not find the xwin ID. In this case, vimtex assumes that the viewer is not started, and so it starts the viewer. Thus you get two instances of zathura. The conclusion is that this is the wrong "track".

Btw: You could also include the content of the patch/diff directly in your post, considering that the length was quite small. In any case, it is no problem to open the patch this way either.

So, back to the issue: I think my pointers [here](https://github.com/lervag/vimtex/issues/1017#issuecomment-349919644) should still be relevant. Basically, it should be possible to pinpoint why the behaviour is different on our setups.

**Comment by jayeshbadwaik** [comment]
So, I have no idea what has happened, but basically with the most recent system update everything works perfectly. I did not change my vimrc at any point, neither did I change any vim related software (zathura, latexmk). Very strange.

**Comment by lervag** [comment]
No problem, I'm happy to hear it works! :)

**Comment by wasaCheney** [comment]
Maybe it is a lillte later, but I encoutner the same problem. 
let g:vimtex_view_automatic=0 could prevent it although not the ideal solution.

PS: freshhand to vim, too hard for me to read .vim files

**Comment by lervag** [comment]
I would not be surprised if this is unrelated and just a consequence of bad configuration on your part. Please open a new issue.
